// ts-to-zod/lib/config.zod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tsToZodConfigSchema = exports.inputOutputMappingSchema = exports.configsSchema = exports.configSchema = exports.customJSDocFormatTypesSchema = exports.customJSDocFormatTypeSchema = exports.customJSDocFormatTypeAttributesSchema = exports.jSDocTagFilterSchema = exports.nameFilterSchema = exports.getSchemaNameSchema = exports.simplifiedJSDocTagSchema = void 0;
// Generated by ts-to-zod
const zod_1 = require("zod");
exports.simplifiedJSDocTagSchema = zod_1.z.object({
    name: zod_1.z.string(),
    value: zod_1.z.string().optional(),
});
exports.getSchemaNameSchema = zod_1.z
    .function()
    .args(zod_1.z.string())
    .returns(zod_1.z.string());
exports.nameFilterSchema = zod_1.z
    .function()
    .args(zod_1.z.string())
    .returns(zod_1.z.boolean());
exports.jSDocTagFilterSchema = zod_1.z
    .function()
    .args(zod_1.z.array(exports.simplifiedJSDocTagSchema))
    .returns(zod_1.z.boolean());
exports.customJSDocFormatTypeAttributesSchema = zod_1.z.object({
    regex: zod_1.z.string(),
    errorMessage: zod_1.z.string().optional(),
});
exports.customJSDocFormatTypeSchema = zod_1.z.string();
exports.customJSDocFormatTypesSchema = zod_1.z.record(exports.customJSDocFormatTypeSchema, zod_1.z.union([zod_1.z.string(), exports.customJSDocFormatTypeAttributesSchema]));
exports.configSchema = zod_1.z.object({
    input: zod_1.z.string(),
    output: zod_1.z.string(),
    skipValidation: zod_1.z.boolean().optional(),
    nameFilter: exports.nameFilterSchema.optional(),
    jsDocTagFilter: exports.jSDocTagFilterSchema.optional(),
    getSchemaName: exports.getSchemaNameSchema.optional(),
    keepComments: zod_1.z.boolean().optional().default(false),
    skipParseJSDoc: zod_1.z.boolean().optional().default(false),
    inferredTypes: zod_1.z.string().optional(),
    customJSDocFormatTypes: exports.customJSDocFormatTypesSchema.optional(),
});
exports.configsSchema = zod_1.z.array(exports.configSchema.and(zod_1.z.object({
    name: zod_1.z.string(),
})));
exports.inputOutputMappingSchema = exports.configSchema.pick({
    input: true,
    output: true,
    getSchemaName: true,
});
exports.tsToZodConfigSchema = zod_1.z.union([exports.configSchema, exports.configsSchema]);
